gem/optimise.py

```diff
@@ -191,6 +196,168 @@ def select_expression(expressions, index):
+    # Pull Delta nodes out of annoying ListTensors, and eliminate
+    # annoying ComponentTensors
+    expression, = remove_componenttensors([expression])
+    expression = pull_delta_from_listtensor(expression)
+    expression, = remove_componenttensors([expression])
+
+    # Flatten a product tree
+    sum_indices = []
+    factors = []
+
+    queue = deque([expression])
+    while queue:
+        expr = queue.popleft()
+        if isinstance(expr, IndexSum):
+            queue.append(expr.children[0])
+            sum_indices.extend(expr.multiindex)
+        elif isinstance(expr, Product):
+            queue.extend(expr.children)
+        else:
+            factors.append(expr)
+
+    # Try to eliminate Delta nodes
+    delta_queue = [(f, index)
+                   for f in factors if isinstance(f, Delta)
+                   for index in (f.i, f.j) if index in sum_indices]
+    while delta_queue:
+        delta, from_ = delta_queue[0]
+        to_, = list({delta.i, delta.j} - {from_})
+
+        sum_indices.remove(from_)
+
+        mapper = MemoizerArg(filtered_replace_indices)
+        factors = [mapper(e, ((from_, to_),)) for e in factors]
+
+        delta_queue = [(f, index)
+                       for f in factors if isinstance(f, Delta)
+                       for index in (f.i, f.j) if index in sum_indices]
+
+    # Drop ones
+    factors = [e for e in factors if e != Literal(1)]
+
+    # Sum factorisation
+    def construct(ordering):
+        """Construct tensor product from a given ordering."""
+        deps = [set(sum_indices) & set(factor.free_indices)
+                for factor in ordering]
+
+        scan_deps = [None] * len(ordering)
+        scan_deps[0] = deps[0]
+        for i in range(1, len(ordering)):
+            scan_deps[i] = scan_deps[i - 1] | deps[i]
+
+        sum_at = [None] * len(ordering)
+        sum_at[0] = scan_deps[0]
+        for i in range(1, len(ordering)):
+            sum_at[i] = scan_deps[i] - scan_deps[i - 1]
+
+        expr = None
+        flops = 0
+        for s, f in reversed(list(zip(sum_at, ordering))):
+            if expr is None:
+                expr = f
+            else:
+                expr = Product(f, expr)
+                flops += numpy.prod([i.extent for i in expr.free_indices], dtype=int)
+            if s:
+                flops += numpy.prod([i.extent for i in s])
+            expr = IndexSum(expr, tuple(i for i in sum_indices if i in s))
+        return expr, flops
+
+    if len(factors) <= 5:
+        expression = None
+        best_flops = numpy.inf
+
+        for ordering in permutations(factors):
+            expr, flops = construct(ordering)
+            if flops < best_flops:
+                expression = expr
+                best_flops = flops
+    else:
+        # Cheap heuristic
+        def key(factor):
+            return len(set(sum_indices) & set(factor.free_indices))
+        ordering = sorted(factors, key=key)
+
+        # FIXME: Log this unexpected case.
```

Comment:
GEM doesn't have a logger, and I don't want to introduce reverse dependency on TSFC.